source/common/http/http1/codec_impl.cc

```diff
@@ -469,8 +469,42 @@ int ConnectionImpl::onHeadersCompleteBase() {
     protocol_ = Protocol::Http10;
   }
   if (Utility::isUpgrade(*current_header_map_)) {
-    ENVOY_CONN_LOG(trace, "codec entering upgrade mode.", connection_);
-    handling_upgrade_ = true;
+    // Ignore h2c upgrade requests until we support them.
+    // See https://github.com/envoyproxy/envoy/issues/7161 for details.
+    if (current_header_map_->Upgrade() &&
+        absl::EqualsIgnoreCase(current_header_map_->Upgrade()->value().getStringView(),
+                               Http::Headers::get().UpgradeValues.H2c)) {
+      ENVOY_CONN_LOG(trace, "removing unsupported h2c upgrade headers.", connection_);
+      current_header_map_->removeUpgrade();
```

Comment:
I'd break this all into a utility function just to keep onHeadersCompleteBase easy to read.  I'm fine if it's in-file though it'd be nifty to make it a generic http utility which takes a vector of string pieces

Reply:
+1 (same file is fine for now)

Reply:
I converted this to a utility function and added it to StringUtil.

Before I write tests could you look at the set of options that I provided and make sure that they are the ones that we should have?

I want to make sure that the signature and options are reasonable before I invest in unnecessary testing for options we don't need.

Reply:
static std::string removeTokens(absl::string_view source, absl::string_view delimiters,
                                  std::set<absl::string_view> tokens_to_remove,
                                  absl::string_view joiner, bool trim_tokens = true,
                                  bool ignore_case = true);
Is what I went with.

Reply:
` static std::string removeTokens(absl::string_view source, absl::string_view delimiters,
                                  std::set<absl::string_view> tokens_to_remove,
                                  absl::string_view joiner, bool trim_tokens = true,
                                  bool ignore_case = true);
`

Reply:
Generally this LGTM. My feeling is to remove any options that you don't need right now. We can always add them later. Thank you!

/wait

Reply:
I have added the tests.  It wasn't hard to keep the options I had so I left them.  It would have been more difficult to explain the default behavior without them.  I also added a benchmark to utility_speed_test which shows that the cost is linear in size.  PTAL

BM_RemoveTokensLong/8                                832 ns          832 ns       845084 bytes_per_second=0/s
BM_RemoveTokensLong/64                              7707 ns         7707 ns        89535 bytes_per_second=0/s
BM_RemoveTokensLong/512                            83653 ns        83654 ns         8406 bytes_per_second=0/s
BM_RemoveTokensLong/4096                          799071 ns       799055 ns          880 bytes_per_second=0/s
BM_RemoveTokensLong/8192                         1696995 ns      1697001 ns          420 bytes_per_second=0/s

Reply:
+1 on remove unused options, by removing ignore_case you can use `CaseUnorderedSet` to avoid `toLower` comparison for the set.